File: boxing/src/main/java/com/bilibili/boxing/model/config/BoxingCropOption.java
/*
 *  Copyright (C) 2017 Bilibili
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.bilibili.boxing.model.config;

import android.net.Uri;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

/**
 * The cropping config, a cropped photo uri is needed at least.
 *
 * @author ChenSL
 */
public class BoxingCropOption implements Parcelable {
    private Uri mDestination;
    private float mAspectRatioX;
    private float mAspectRatioY;
    private int mMaxWidth;
    private int mMaxHeight;

    public BoxingCropOption(Uri destination) {
        this.mDestination = destination;
    }

    public static BoxingCropOption with(@NonNull Uri destination) {
        return new BoxingCropOption(destination);
    }

    public BoxingCropOption aspectRatio(float x, float y) {
        this.mAspectRatioX = x;
        this.mAspectRatioY = y;
        return this;
    }

    public BoxingCropOption useSourceImageAspectRatio() {
        this.mAspectRatioX = 0;
        this.mAspectRatioY = 0;
        return this;
    }

    public BoxingCropOption withMaxResultSize(int width, int height) {
        this.mMaxWidth = width;
        this.mMaxHeight = height;
        return this;
    }


    public float getAspectRatioX() {
        return mAspectRatioX;
    }

    public float getAspectRatioY() {
        return mAspectRatioY;
    }

    public int getMaxHeight() {
        return mMaxHeight;
    }

    public int getMaxWidth() {
        return mMaxWidth;
    }

    public Uri getDestination() {
        return mDestination;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(this.mDestination, flags);
        dest.writeFloat(this.mAspectRatioX);
        dest.writeFloat(this.mAspectRatioY);
        dest.writeInt(this.mMaxWidth);
        dest.writeInt(this.mMaxHeight);
    }

    BoxingCropOption(Parcel in) {
        this.mDestination = in.readParcelable(Uri.class.getClassLoader());
        this.mAspectRatioX = in.readFloat();
        this.mAspectRatioY = in.readFloat();
        this.mMaxWidth = in.readInt();
        this.mMaxHeight = in.readInt();
    }

    public static final Creator<BoxingCropOption> CREATOR = new Creator<BoxingCropOption>() {
        @Override
        public BoxingCropOption createFromParcel(Parcel source) {
            return new BoxingCropOption(source);
        }

        @Override
        public BoxingCropOption[] newArray(int size) {
            return new BoxingCropOption[size];
        }
    };
}

File: boxing/src/main/java/com/bilibili/boxing/model/config/BoxingCropOption.kt
/*
 *  Copyright (C) 2017 Bilibili
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.bilibili.boxing.model.config

import android.net.Uri
import android.os.Parcel
import android.os.Parcelable

/**
 * The cropping config, a cropped photo uri is needed at least.

 * @author ChenSL
 */
class BoxingCropOption : Parcelable {
    var destination: Uri? = null
        private set
    var aspectRatioX: Float = 0.toFloat()
        private set
    var aspectRatioY: Float = 0.toFloat()
        private set
    var maxWidth: Int = 0
        private set
    var maxHeight: Int = 0
        private set

    constructor(destination: Uri) {
        this.destination = destination
    }

    fun aspectRatio(x: Float, y: Float): BoxingCropOption {
        this.aspectRatioX = x
        this.aspectRatioY = y
        return this
    }

    fun useSourceImageAspectRatio(): BoxingCropOption {
        this.aspectRatioX = 0f
        this.aspectRatioY = 0f
        return this
    }

    fun withMaxResultSize(width: Int, height: Int): BoxingCropOption {
        this.maxWidth = width
        this.maxHeight = height
        return this
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(this.destination, flags)
        dest.writeFloat(this.aspectRatioX)
        dest.writeFloat(this.aspectRatioY)
        dest.writeInt(this.maxWidth)
        dest.writeInt(this.maxHeight)
    }

    internal constructor(`in`: Parcel) {
        this.destination = `in`.readParcelable<Uri>(Uri::class.java.classLoader)
        this.aspectRatioX = `in`.readFloat()
        this.aspectRatioY = `in`.readFloat()
        this.maxWidth = `in`.readInt()
        this.maxHeight = `in`.readInt()
    }

    companion object {

        fun with(destination: Uri): BoxingCropOption {
            return BoxingCropOption(destination)
        }

        val CREATOR: Parcelable.Creator<BoxingCropOption> = object : Parcelable.Creator<BoxingCropOption> {
            override fun createFromParcel(source: Parcel): BoxingCropOption {
                return BoxingCropOption(source)
            }

            override fun newArray(size: Int): Array<BoxingCropOption?> {
                return arrayOfNulls(size)
            }
        }
    }
}